connectors/sources/s3.py

```diff
@@ -35,57 +37,128 @@
         if AWS_ENDPOINT is not None:
             logger.debug(f"Creating a session against {AWS_ENDPOINT}")
 
-        async with self.session.client(
-            service_name="s3", config=self.config, endpoint_url=AWS_ENDPOINT, **kwargs
+        # AsyncExitStack, supports asynchronous context managers, used to create client using enter_async_context and
+        # these context manager will be stored in client_context list also client will be stored in clients dict with their region
+        s3_context_stack = AsyncExitStack()
+        s3_client = await s3_context_stack.enter_async_context(
+            self.session.client(
+                service_name="s3",
+                config=self.config,
+                endpoint_url=AWS_ENDPOINT,
+                region_name=region,
+            )
+        )
+        self.client_context.append(s3_context_stack)
+
+        self.clients[region_name] = s3_client
+        return self.clients[region_name]
+
+    async def close_client(self):
+        """Closes unclosed client session"""
+        for context in self.client_context:
+            await context.aclose()
+
+    async def fetch_buckets(self):
+        """This method used to list all the buckets from Amazon S3"""
+        s3 = await self.client()
+        await s3.list_buckets()
+
+    async def get_bucket_list(self):
+        """Returns bucket list from list_buckets response
+
+        Returns:
+            list: List of buckets
+        """
+        if self.configuration["buckets"] == ["*"]:
+            s3 = await self.client()
+            bucket_list = await s3.list_buckets()
+            buckets = [bucket["Name"] for bucket in bucket_list["Buckets"]]
+        else:
+            buckets = self.configuration["buckets"]
+        return buckets
+
+    async def get_bucket_objects(self, bucket):
+        """Returns bucket list from list_buckets response
+        Args:
+            bucket (str): Name of bucket
+        Yields:
+            obj_summary: Bucket objects metadata
+            s3_client: S3 client object
+        """
+        page_size = self.configuration["page_size"]
+        region_name = await self.get_bucket_region(bucket)
+        s3_client = await self.client(region=region_name)
+        async with self.session.resource(
+            service_name="s3",
+            config=self.config,
+            endpoint_url=AWS_ENDPOINT,
+            region_name=region_name,
         ) as s3:
-            yield s3
+            try:
+                bucket_obj = await s3.Bucket(bucket)
+                await asyncio.sleep(0)
 
-    async def ping(self):
-        """Verify the connection with AWS"""
+                async for obj_summary in bucket_obj.objects.page_size(page_size):
+                    yield obj_summary, s3_client
+            except Exception as exception:
+                logger.warning(
+                    f"Something went wrong while fetching documents from {bucket}. Error: {exception}"
```

Comment:
I _think_ that instead of `logger.warning` it should be `log_warn` now.

Reply:
I am not completely sure about log_warn but logging.warn seems deprecated now.
https://python.plainenglish.io/logging-warning-vs-logging-warn-in-python-c0b4b927515#:~:text=Until%20Python%203.3%2C%20logging.warn()%20and%20logging.warning()%20were%20exactly%20the%20same%20with%20the%20only%20difference%20being%20that%20the%20former%20remained%20undocumented.%20They%20were%20literally%20pointing%20to%20the%20same%20object%3A

Reply:
yeah `logger.warning` is correct.`warn` is deprecated

Reply:
I'm referencing work that @wangch079 has been doing to improve the metadata included in all our log messages. @wangch079 can you confirm that `log_warn`, `log_info` etc are what we should be using in connectors going forward, and share where the definitions of those functions are?

Reply:
Yes, you should use `logger.warning` instead of `logger.warn`.

> @wangch079 can you confirm that `log_warn`, `log_info` etc are what we should be using in connectors going forward

It's `log_warning`:
https://github.com/elastic/connectors-python/blob/ebc2e8dd83b52dc4db8d41511825e52fbc71b49d/connectors/es/document.py#L71-L78

But the log correlation doesn't apply to each connector implementation yet, which is worth a separate discussion in PR https://github.com/elastic/connectors-python/pull/806.

For this PR, I think it's good to keep using `logger.warning`.